### DISCUSSION TITLE
Ignore crate: is it expected that `WalkBuilder#types` overrides `hidden` default?

### DISCUSSION THREAD

**Discussion by nwj** [BurntSushi/ripgrep#1699] [Q&A] [answered]
Hello 😄 ,

I have code like this using v0.4.16 of the ignore crate:
```rust
let matcher = TypesBuilder::new().add_defaults().select("markdown").build().unwrap();
for entry in WalkBuilder::new("./").types(matcher).build() {
  # stuff happens...
}
```

My expectation was that this wouldn't match hidden markdown files (e.g. something like `.hidden-file.md`) but actually it does. Is that expected behavior? 

What I think is going on is that the `types` call matches any entry with a markdown file extension, even if the entry is hidden and the walk is configured to skip hidden files. I believe this because when I remove `.types(matcher)` from the above, the hidden markdown file is no longer matched.

**Comment by BurntSushi** [comment, marked as answer]
Yes, it's correct. The intent here is that you've explicitly given a type you want to search, and that should generally override more general default behavior. If you want to search non-hidden markdown files, you should be able to filter out hidden files pretty easily. Or you could add an override glob that does it, e.g., `![!.]*` or something like that.

Technically, this is somewhat covered by the [rule precedence in the docs](https://docs.rs/ignore/0.4.16/ignore/struct.WalkBuilder.html#ignore-rules), although it can be a bit hard to interpret.

Specifically, the third rule states:

> Third, if the previous step yields an ignore match, then all matching is stopped and the path is skipped. If it yields a whitelist match, then matching continues. A whitelist match can be overridden by a later matcher.

Now, a "type" matcher isn't explicitly mentioned in rules 1 or 2, but I think it comes after rule 1 but before rule 2. And then the fifth rule says:

> Fifth, if the path hasn't been whitelisted and it is hidden, then the path is skipped.

Which is basically saying, "unless something else has matched it, as a last ditch attempt to filter stuff out, ignore it if it's hidden." Overall that means that any kind of explicit whitelist rule will override the hidden filtering.

**Reply by nwj** [reply]
Thanks for the clarification!
